Add guideline for using safe operator. (#912)
rubocop/ruby-style-guide: README.adoc

foo&.bar
----

> === Safe navigation

Reviewer: I'd argue regarding the section, Source code layout.
Maybe it would be a better fit for [Misc](https://rubystyle.guide/#misc)?

Author: Can you elaborate more regarding this comment?

Reviewer: If I'm not mistaken, presently this new guideline resides in the "Source code layout" section. This section is mostly all about whitespace, formatting, those things that don't change code semantics.

The new guideline suggests code changes that may, depending on constraints, change its behaviour.

Basing on this observation, I propose moving this new guideline to a different section. I don't have a better section in mind, so I suggested "Misc".

Author: got it

> Avoid long chains of `&.`. Replace with `.` if they introduce unnecessary conditional logic. 

Reviewer: "long" sounds vague. How do you feel about removing it altogether?

Author: yes sure what alternative would you suggest after removing it?

> Avoid long chains of `&.`. Replace with `.` if they introduce unnecessary conditional logic. 

Reviewer: What does it mean "if they introduce unnecessary conditional logic"?

Author: Updated the sentence, I mean to say if introducing the safe operator adds unnecessary conditional logic we should use other methods such as delegation.

Consider other approaches such as delegation.

> For condition where users are guaranteed to have an address and addresses are guaranteed to have a zip code

Reviewer: ```suggestion
If users are guaranteed to have an address and addresses are guaranteed to have a zip code:
```

> [source, thoughtbot(https://thoughtbot.com/blog/ruby-safe-navigation#conditional-equivalence)]

Reviewer: This breaks rendering
![image](https://user-images.githubusercontent.com/6916/209480000-9b13d0e4-c009-4a8a-ba77-1093ccee9904.png)

foo&.bar
----

=== Safe navigation

> Avoid long chains of `&.` replace with `.`, if safe navigator operator introduce unnecessary conditional logic.
> Consider other approaches such as delegation.
> 
> If users are guaranteed to have an address and addresses are guaranteed to have a zip code:

Reviewer: ```suggestion
Avoid chaining of `&.`. Replace with `.` and an explicit check.
E.g. if users are guaranteed to have an address and addresses are guaranteed to have a zip code:
```

> [source, thoughtbot]

Reviewer: ```suggestion
[source,ruby]
```
According to the [AsciiDoc documentation](https://docs.asciidoctor.org/asciidoc/latest/verbatim/source-blocks/), the second argument is the language of the source.

No wonder "thoughtbot" here breaks code highlighting:
![image](https://user-images.githubusercontent.com/6916/209580738-28d81f91-5433-4402-a090-9fecb7f5b09e.png)

----
# bad
user&.address&.zip

#good 
user && user.address.zip

> 
> #good
> class User
>   def zip
>     address.zip
>   end
> end
> 
> user&.zip

Reviewer: ```suggestion
----

If such a change introduces excessive conditional logic, consider other approaches, such as delegation:
[source]
----
#bad
user && user.address && user.address.zip

#good
class User
  def zip
    address&.zip
  end
end
user&.zip
```
